internal/compare/compare.go

```diff
@@ -0,0 +1,177 @@
+}
+
+func (e changOptToReqError) Error() string {
+	return fmt.Sprintf("changing an optional field %s in %s to required is not backwards compatible",
+		e.field, e.struc)
+}
+
+type removeMethodError struct {
+	method  string
+	service string
+}
+
+func (e removeMethodError) Error() string {
+	return fmt.Sprintf("removing method %s in service %s is not backwards compatible", e.method, e.service)
+}
+
+type deleteServiceError struct {
+	service string
+}
+
+func (e deleteServiceError) Error() string {
+	return fmt.Sprintf("deleting service %s is not backwards compatible", e.service)
+}
+
+// Modules looks for removed methods and added required fields.
+func (p *Pass) Modules(fromModule, toModule *compile.Module) {
+	p.services(fromModule, toModule)
+	p.checkRequiredFields(fromModule, toModule)
+}
+
+// Diagnostic is a message associated with an error and a file name.
+type Diagnostic struct {
+	File string // File where error was discovered
+	Err  error  // Specific error
+}
+
+func (d *Diagnostic) String() string {
+	return fmt.Sprintf("file: %s, error: %s", d.File, d.Err)
+}
+
+// Pass provides all reported errors.
+type Pass struct {
+	lints []Diagnostic
+}
+
+// Report reports an error.
+func (p *Pass) Report(d Diagnostic) {
+	p.lints = append(p.lints, d)
+}
+
+// Lints returns all errors.
+func (p *Pass) Lints() []Diagnostic {
+	return p.lints
+}
+
+func (p *Pass) String() string {
+	var s string
+	for _, l := range p.lints {
+		s += fmt.Sprintf("%s\n", l.String())
+	}
+
+	return s
+}
+
+func (p *Pass) checkRequiredFields(fromModule, toModule *compile.Module) {
+	for n, spec := range toModule.Types {
+		fromSpec, ok := fromModule.Types[n]
+		if !ok {
+			// This is a new Type, which is backwards compatible.
+			continue
+		}
+		if s, ok := spec.(*compile.StructSpec); ok {
+			// Match on Type names. Here we hit a limitation, that if someone
+			// renames the struct and then adds a new field, we don't really have
+			// a good way of tracking it.
+			if fromStructSpec, ok := fromSpec.(*compile.StructSpec); ok {
+				p.structSpecs(fromStructSpec, s, filepath.Base(fromModule.ThriftPath))
+			}
+		}
+	}
+}
+
+// StructSpecs compares two structs defined in a Thrift file.
+func (p *Pass) structSpecs(from, to *compile.StructSpec, file string) {
+	fields := make(map[int16]*compile.FieldSpec, len(from.Fields))
+	// Assume that these two should be compared.
+	for _, f := range from.Fields {
+		// Capture state of all fields here.
+		fields[f.ID] = f
+	}
+
+	for _, toField := range to.Fields {
+		if fromField, ok := fields[toField.ID]; ok {
+			fromRequired := fromField.Required
+			toRequired := toField.Required
+			if !fromRequired && toRequired {
+				p.Report(Diagnostic{
+					File: file,
+					Err: changOptToReqError{
+						field: toField.ThriftName(),
+						struc: to.ThriftName(),
+					},
+				})
+			}
+		} else if toField.Required {
+			p.Report(Diagnostic{
+				File: file,
+				Err:  addReqError{toField.ThriftName(), to.ThriftName()},
```

Comment:
Another reason to not use `error` objects for these: readabilitiy.
I have no way of knowing what these issues are without going to the error definition.